tests/federation_room_join_partial_state_test.go

```diff
@@ -3452,35 +3452,33 @@ func TestPartialStateJoin(t *testing.T) {
 		})
 		terry.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "directory", "list", "room", serverRoom.RoomID}, reqBody)
 
+		assertPublicRoomDirectoryMemberCountEquals := func(t *testing.T, expectedMemberCount int64) {
+			// In Synapse, rooms stats are updated by a background job which runs asynchronously.
+			// To account for that, we check for up to 3 seconds that the job has completed.
+			terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"},
+				client.WithRetryUntil(time.Second*3, func(res *http.Response) bool {
+					body, err := ioutil.ReadAll(res.Body)
+					if err != nil {
+						t.Fatalf("something broke: %v", err)
+					}
+					numJoinedMembers := gjson.GetBytes(body, "chunk.0.num_joined_members")
+					if numJoinedMembers.Int() == expectedMemberCount {
+						return true
+					}
+					return false
+				}))
+		}
+
 		// sanity check - before the state has completed syncing state we would expect only one user
 		// to show up in the room list
-		res := terry.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
-
-		must.MatchResponse(t, res, match.HTTPResponse{
-			StatusCode: 200,
-			JSON: []match.JSON{
-				match.JSONKeyEqual("chunk.0.num_joined_members", 1),
-			}})
+		assertPublicRoomDirectoryMemberCountEquals(t, 1)
```

Comment:
To check: this PR adds retry logic to the sanity check here?

Reply:
Yes, it adds retry logic and then factors it out into a function.